components/parquet_ext/src/async_arrow_writer.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+// Copyright 2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+//! Async arrow writer for parquet file.
+
+use std::{
+    io::Write,
+    sync::{Arc, Mutex},
+};
+
+use arrow::{datatypes::SchemaRef, record_batch::RecordBatch};
+use parquet::{
+    arrow::ArrowWriter,
+    errors::{ParquetError, Result},
+    file::properties::WriterProperties,
+    format::{FileMetaData, KeyValue},
+};
+use tokio::io::{AsyncWrite, AsyncWriteExt};
+
+#[derive(Clone, Default)]
+pub struct SharedBuffer {
+    /// The buffer to store the data to be written.
+    ///
+    /// The lock is used to obtain internal mutability, so no worry about the
+    /// lock contention.
+    buffer: Arc<Mutex<Vec<u8>>>,
+}
+
+impl Write for SharedBuffer {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let mut buffer = self.buffer.lock().unwrap();
+        Write::write(&mut *buffer, buf)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        let mut buffer = self.buffer.lock().unwrap();
+        Write::flush(&mut *buffer)
+    }
+}
+
+/// Async arrow writer for parquet file with a buffer whose size limit is
+/// configurable.
+pub struct AsyncArrowWriter<W> {
+    sync_writer: ArrowWriter<SharedBuffer>,
+    async_writer: W,
+    shared_buffer: SharedBuffer,
+    max_buffer_size: usize,
+}
+
+impl<W: AsyncWrite + Unpin + Send> AsyncArrowWriter<W> {
+    pub fn try_new(
+        writer: W,
+        arrow_schema: SchemaRef,
+        max_buffer_size: usize,
+        props: Option<WriterProperties>,
+    ) -> Result<Self> {
+        let shared_buffer = SharedBuffer::default();
+        let sync_writer = ArrowWriter::try_new(shared_buffer.clone(), arrow_schema, props)?;
+
+        Ok(Self {
+            sync_writer,
+            async_writer: writer,
+            shared_buffer,
+            max_buffer_size,
+        })
+    }
+
+    pub async fn write(&mut self, batch: &RecordBatch) -> Result<()> {
+        self.sync_writer.write(batch)?;
+        Self::flush(
+            &self.shared_buffer,
+            &mut self.async_writer,
+            self.max_buffer_size,
+        )
+        .await
+    }
+
+    pub fn append_key_value_metadata(&mut self, kv_metadata: KeyValue) {
+        self.sync_writer.append_key_value_metadata(kv_metadata);
+    }
+
+    pub async fn close(mut self) -> Result<FileMetaData> {
+        let metadata = self.sync_writer.close()?;
+
+        // flush the remaining data.
+        Self::flush(&self.shared_buffer, &mut self.async_writer, 0).await?;
+        self.async_writer
+            .shutdown()
+            .await
+            .map_err(|e| ParquetError::External(Box::new(e)))?;
+
+        Ok(metadata)
+    }
+
+    async fn flush(
+        shared_buffer: &SharedBuffer,
```

Comment:
Why this flush is not an associated method? `fn flush(&mut self ...)`

Reply:
When implementing `close` method, the `sync_writer` has to be consumed before `flush` which requires the `flush` method not to be the associated method.